### PR TITLE
Split test jobs in Travis

### PR DESCRIPTION
**Pull request by chetanmeh**
Splits the Travis build to have 2 separate jobs for running tests

## Description

This PR splits the travis setup in 2 parts 

1. Test requiring only database
2. Test requiring complete OpenWhisk system

See discussions in #3572 for details

## Implementation Details

This PR introduces a notion of `testSets` which define set of filters for which test to execute. It defines 3 sets for now

1. `REQUIRE_ONLY_DB` - Test which just need database to be up
2. `REQUIRE_SYSTEM` - Test which require whole OpenWhisk runtime up. Excludes tests under 1
3. `LEAN` - Include all lean test i.e. current master

### Changes in checkLogs.py

`checkLogs.py` currently performs some log sanity check like checking if db, controller logs are non empty. With this PR for `REQUIRE_ONLY_DB` controller logs would not be there. So it has been modified to support `tags`. Each check specify set of tags to which it belongs like `db` and `system`. And then `checkLogs.py` would specify the tags which it wants to enforce checks for

### TODO

- [x] - Optimize jobs to remove steps not required. For e.g. avoid deploying containers for pure db tests

## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#3572)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by chetanmeh** [comment]
image

Did few runs on my fork and have following observation

1. Having multiple jobs seems to slow down task compared to when there were only 2 jobs. For e.g. the wrk-performance jobs used to take ~14-15 mins earlier and now it took 20 mins

As Apache has a pool of [30 instances][1] would see how much time it takes as part of PR build which uses same pool

[1]: https://blogs.apache.org/infra/entry/apache_gains_additional_travis_ci

**Comment by chetanmeh** [comment]
image

Current build timings for this PR. Build time reduces from ~44 mins to ~35 mins. Next would try splitting and optimizing the build steps

**Comment by chetanmeh** [comment]
Post splitting of `build.sh` time has come down to ~ 31 mins

image

File: .travis.yml [hunk 1]
```diff
@@ -48,18 +48,32 @@ notifications:
...
-        - ./tools/travis/build.sh
+        - export ORG_GRADLE_PROJECT_testSetName="REQUIRE_ONLY_DB"
+        - ./tools/travis/scan.sh
+        - ./tools/travis/setupPrereq.sh
+        - ./tools/travis/distDocker.sh
+        - ./tools/travis/runTests.sh
         - ansible-playbook -i ansible/environments/local ansible/logs.yml # collect logs regardless of build exit
-        - ./tools/travis/box-upload.py "$TRAVIS_BUILD_DIR/logs" "$TRAVIS_BUILD_ID-$TRAVIS_BRANCH.tar.gz"
-      env: DESCRIPTION="Execute test suite."
+        - tools/build/checkLogs.py logs db
```
Comment (csantanapr) [inline .travis.yml]: Not consistent `./tools` starting with `./`

File: tests/build.gradle [hunk 2]
```diff
@@ -33,8 +34,70 @@ def leanExcludes = [
...
+testSets.each {setName, patterns ->
+    def excludes = patterns["excludes"] ?: new HashSet<>()
+    excludes.addAll(leanExcludes)
+    patterns["excludes"] = excludes
+}
+
+//The value can be specified either via env variable
+// ORG_GRADLE_PROJECT_testSetName
+//Or via property -PtestSetName
+if (!project.hasProperty("testSetName")) {
```
[the remark this answers is not in the document]
Comment (chetanmeh) [reply tests/build.gradle]: This was a misleading log which was getting logged due to build task evaluation triggering print message to be logged. Despite this log all tests would have been run.

This has been fixed now

**Comment by chetanmeh** [comment]
Added a new task to just run unit tests which do not require even a running CouchDB server locally. 

    $ ansible-playbook -i ansible/environments/local ansible/properties.yml
    $ ./gradlew tests:testUnit

This task runs 823 tests in 2.5 mins!. So developers can run a large portion of test suite locally and can test changes locally for early feedback and faster development

**Comment by csantanapr** [comment]
I tested this internally and `.gradlew tests:test` still runs all tests

**Pull request merged**